### PR TITLE
chore: Setup `tslib` & `importHelpers` to improve bundle size

### PR DESCRIPTION
**Pull request by jimblanc**
#### Description of changes
This change enables `importHelpers` in our shared build script (`build.js`) and adds `tslib` as a dependency to all packages that use the script to improve bundle size.

##### Note on methodology 
I collected the data below by looking at the minified WebPack bundle files that each of our modules produce. While this may not reflect all use-cases, it provides a readily available & consistent metric to evaluate the impact of the change. As far as dependency de-duping & flattening for more complex apps, the numbers below _should_ capture the worst-case where each package includes the minimum number of `tslib` dependencies that it requires without any additional de-duping that the importing app may perform.

| Package | Artifact | Before Change | After Change | Bundle Δ |
| --- | --- | --- | --- | --- |
| `aws-amplify` | aws-amplify.min.js | 1294 KiB | 1294 KiB | 0 KiB |
| `analytics` | aws-amplify-analytics.min.js | 296 KiB | 285 KiB | -11 KiB |
| `api` | aws-amplify-api.min.js | 146 KiB | 145 KiB | -1 KiB |
| `api-graphql` | aws-amplify-api-graphql.min.js | 144 KiB | 142 KiB | -2 KiB |
| `api-rest` | aws-amplify-api-rest.min.js | 43 KiB | 41 KiB | -2 KiB |
| `auth` | aws-amplify-auth.min.js | 157 KiB | 155 KiB | -2 KiB |
| `cache` | aws-amplify-cache.min.js  | 15.5 KiB | 15.7 KiB | +0.2 KiB |
| `core` | aws-amplify-core.min.js  | 290 KiB | 281 KiB | -9 KiB |
| `datastore` | aws-amplify-datastore.min.js  | 617 KiB | 590 KiB | -27 KiB |
| `datastore-storage-adapter` | aws-amplify-datastore-sqlite-adapter-expo.min.js, aws-amplify-datastore-storage-adapter.min.js  | 47.6 KiB | 44.9 KiB | -2.7 KiB |
| `geo` | aws-amplify-geo.min.js  | 218 KiB | 216 KiB | -2 KiB |
| `interactions` | aws-amplify-interactions.min.js  | 323 KiB | 315 KiB | -8 KiB |
| `predictions` | aws-amplify-predictions.min.js  | 556 KiB | 556 KiB | 0 KiB |
| `storage` | aws-amplify-storage.min.js | 328 KiB | 322 KiB | -6 KiB |
| `xr` | aws-amplify-xr.min.js | 14.7 KiB | 13.6 KiB | -1.1 KiB |

**Total Δ for entire project** -73.6 KiB

**Note** Soon-to-be deprecated UI packages, as well as the `pushnotification` package (which is not used for web & does not use the shared build script) were not included in this change.

##### Future improvements
- Refactor `build.js` so that each package can have its own `tsconfig.json`
  - This will allow us to disable `importHelpers` for packages where it doesn't provide any benefit such as `cache`
- Upgrade packages that depend on older (`< 2.0.0`) versions of `tslib` to improve dependency flattening, for example:
  - `@aws-crypto/sha256-js`
  - `@aws-sdk/credential-provider-cognito-identity`
  - `@aws-sdk/client-cognito-identity`

#### Issue #, if available
NA

#### Description of how you validated changes
- Local release build & unit test pass
- Manually evaluated WebPack'd bundle sizes

Integ test pass on `next-major-version/5`: https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/15155/workflows/3a454e6b-c4e8-47aa-9190-359b225d4ffd
  - Existing storage failures already present
 
#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/__tests__/hosted-ui.test.ts [hunk 1]
```diff
@@ -160,6 +160,11 @@ jest.mock('amazon-cognito-identity-js/lib/CognitoUser', () => {
 import * as AmplifyCore from '@aws-amplify/core';
 const { Hub, Credentials, StorageHelper } = AmplifyCore;
 
+// Mock the module to ensure that setters are available for spying
+jest.mock('@aws-amplify/core', () => ({
+	...jest.requireActual('@aws-amplify/core'),
+}));
+
```
Comment (manueliglesias) [inline packages/auth/__tests__/hosted-ui.test.ts:167]: Is this related to `tslib`?

**Review by manueliglesias — APPROVED** [review]
LGTM!

**Review by AllanZhengYP — APPROVED** [review]

File: packages/auth/__tests__/hosted-ui.test.ts (hunk 1 above)
Comment (jimblanc) [reply packages/auth/__tests__/hosted-ui.test.ts:167]: So from what I can tell this is kind of a known issue with Jest & TypeScript because babel will only include getters for re-exported functions (like `browserOrNode` in this file). I'm not entirely sure why this was working before though. I double checked the TS configuration and experimented with setting `importHelpers` through the `ts-jest` configuration but it didn't seem to affect anything one way or the other.
